source/slang/slang-emit-c-like.cpp

```diff
@@ -1657,11 +1672,16 @@ bool CLikeSourceEmitter::shouldFoldInstIntoUseSites(IRInst* inst)
     return true;
 }
 
+bool CLikeSourceEmitter::isPointerSyntaxRequiredImpl(IRInst* /* inst */)
```

Comment:
Why is this needed?

Reply:
At first I just flipped `doesTargetSupportPtrTypes` to `true` for WGSL, but that caused a lot of strange output that didn't make sense (or compile), like e.g:

    *(&array)[i] = 5i; // instead of 'array[i] = 5i'
    
Therefore, I added `isPointerSyntaxRequiredImpl` (defaulting to `doesTargetSupportPtrTypes`) and replaced some calls to `doesTargetSupportPtrTypes` by `isPointerSyntaxRequiredImpl` which also takes the `IRInst` as context in order to be able to opt-in to the "pointer syntax path" in some cases.

Reply:
My expectation is that `doesTargetSupportPtrTypes` needs to be false for WGSL backend. What are the cases where we want to emit pointer syntax in wgsl?

Reply:
Here is the function, with the various cases shere I opt out of the "pointer syntax path":


    bool WGSLSourceEmitter::isPointerSyntaxRequiredImpl(IRInst* inst)
    {
        // Case 1: Structured buffers are mapped to 'array' types, which don't need dereferencing
        if (inst->getOp() == kIROp_RWStructuredBufferGetElementPtr)
            return false;
    
        // Case 2: Don't emit "->" to access fields in resource structs
        if (inst->getOp() == kIROp_FieldAddress)
            return false;
    
        // Case 3: Don't emit "*" to access fields in resource structs
        if (inst->getOp() == kIROp_GlobalParam)
            return false;
    
        // Case 4: Emit 'globalVar' instead of  "*&globalVar"
        if (inst->getOp() == kIROp_GlobalVar)
            return false;
    
        return true;
    }


Here is an example of what goes wrong if I don't opt out in each of the cases:

*Case 1: (tests/compute/array-param.slang)*

    @binding(0) @group(0) var<storage, read_write> globalParams_outputBuffer_0 : array<i32>;

    // ...

    @compute
    @workgroup_size(4, 1, 1)
    fn computeMain(@builtin(global_invocation_id) dispatchThreadID_0 : vec3<u32>)
    {
        var dispatchThreadID_1 : vec3<i32> = vec3<i32>(dispatchThreadID_0);
        var b_0 : array<vec3<f32>, i32(4)>;
        writeArray_0(&b_0);
    // error: wgpu: type: 2: message: Error while parsing WGSL: :20:7 error: cannot dereference expression of type 'i32'
        *(&globalParams_outputBuffer_0)[dispatchThreadID_1.x] = i32(b_0[i32(0)].x);
    //    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    // When opting out of "pointer syntax path", I get valid WGSL:
    //  (&globalParams_outputBuffer_0)[dispatchThreadID_1.x] = i32(b_0[i32(0)].x);
    
        
        return;
    }

*Case 2: (tests/compute/cast-zero-to-struct.slang)*

    @binding(0) @group(0) var<storage, read_write> globalParams_gOutputBuffer_0 : array<i32>;
    
    struct S_0
    {
         a_0 : vec2<i32>,
         b_0 : vec2<i32>,
         c_0 : vec2<i32>,
    };
    
    fn test_0( val_0 : i32) -> i32
    {
        var _S1 : vec2<i32> = vec2<i32>(i32(0), i32(0));
        var s_0 : S_0;
    // error: wgpu: type: 2: message: Error while parsing WGSL: :14:11 error: expected '=' for assignment
        (&s_0)->a_0 = _S1;
    //        ^^
    // When opting out of "pointer syntax path", I get valid WGSL:
    //  (&s_0).a_0 = _S1;

        // ...
    }

*Case 3: (tests/compute/dynamic-dispatch-11.slang)*


    struct Tuple_std140_0
    {
        @align(16) value0_0 : vec2<u32>,
        @align(8) value1_0 : vec2<u32>,
        @align(16) value2_0 : _std140_0,
    };
    
    @binding(16) @group(0) var<uniform> globalParams_gCb_0 : Tuple_std140_0;
    
    fn _S7( _S8 : i32) -> i32
    {
    // error: wgpu: type: 2: message: Error while parsing WGSL: :61:60 error: cannot dereference expression of type 'Tuple_std140_0'
        return MyImpl_run_0(unpackAnyValue8_0(unpackStorage_1(*globalParams_gCb_0).value2_0), _S8);
    //                                                         ^^^^^^^^^^^^^^^^^^
    // When opting out of "pointer syntax path", I get valid WGSL:
    //  return MyImpl_run_0(unpackAnyValue8_0(unpackStorage_1(globalParams_gCb_0).value2_0), _S8);    
    }

*Case 4: (tests/compute/global-init.slang)*

    @binding(0) @group(0) var<storage, read_write> globalParams_outputBuffer_0 : array<i32>;
    
    const gVar_0 : i32 = i32(16);
    
    fn test_0( inVal_0 : i32) -> i32
    {
    // error: wgpu: type: 2: message: Error while parsing WGSL: :7:24 error: cannot take the address of 'const gVar_0'
        return inVal_0 + *&gVar_0;
    //                     ^^^^^^
    // When opting out of "pointer syntax path", I get valid WGSL:
    //  return inVal_0 + gVar_0;
    }
    
    @compute
    @workgroup_size(4, 1, 1)
    fn computeMain(@builtin(global_invocation_id) dispatchThreadID_0 : vec3<u32>)
    {
        var tid_0 : u32 = dispatchThreadID_0.x;
        (&globalParams_outputBuffer_0)[tid_0] = test_0((&globalParams_outputBuffer_0)[tid_0]);
        return;
    }

Reply:
I agree this workaround feels wrong, but I'm not sure how to do it right.
It seems I do need the "pointer syntax path" in many cases, but I was pretty confused about the output in the above cases, which I saw when I just set `doesTargetSupportPtrTypes` to return `true` unconditionally.

Reply:
If I have `int arr[3]` in c++, `*&arr[0]` is valid code. I means take address of `arr`, then dereference that address to get a reference to the array, and then index into the array. It seems that the problem is in wgsl, you can't just do `arr[i]`, and have to do `&arr[i]`, which is weird from C/C++ perspective.

Reply:
I guess my original question was the other way around: why can't the existing behavior of doesTargetSupportPtrTypes=false work out of the box? Why do we need to customize the behavior?

Reply:
Why do you need to do `(&globalParams_outputBuffer_0)[tid_0]`? `globalParams_outputBuffer_0` is already an array, and you shouldn't need to take its address before access its elements.

Reply:
I think we shouldn't emit things like `(&arr)[i]`, because `(&arr)`'s type is `ptr<array<int>>`, and it is not clear to me what it means to use `[]` affer an `ptr<array<int>>`, if this is c++, then the result of that is `ref<array<int>>`, which shouldn't allow assignment from an int to it.

Reply:
If we delete `isPointerSyntaxRequiredImpl` and just use the default `doesTargetSupportPointer = false` behavior, what will break?

Reply:
"so arr[i] is not an l-value and I can't pass it to a function that expects inout parameter?"

I guess you mean just 'arr'? If so then yes I think that's correct. If you're passing it to a function, the function needs to take a `ptr<arr>` and you need to pass `&arr` to the function.

Reply:
> If we delete `isPointerSyntaxRequiredImpl` and just use the default `doesTargetSupportPointer = false` behavior, what will break?

Categories of failures when I never take the "pointer syntax" paths:

1. Passing function parameter without &
2. Assigning to ptr without *
3. Reading T from ptr-to-T without *

Here are examples:

*Category 1 (tests/compute/array-param)*

    fn writeArray_0( a_0 : ptr<function, array<vec3<f32>, i32(4)>>)
    {
        // ...
    }

    @compute
    @workgroup_size(4, 1, 1)
    fn computeMain(@builtin(global_invocation_id) dispatchThreadID_0 : vec3<u32>)
    {
        var dispatchThreadID_1 : vec3<i32> = vec3<i32>(dispatchThreadID_0);
        var b_0 : array<vec3<f32>, i32(4)>;
    // Error while parsing WGSL: :19:18 error: type mismatch for argument 1 in call to 'writeArray_0',
    //   expected 'ptr<function, array<vec3<f32>, 4>, read_write>',
    //   got 'array<vec3<f32>, 4>'        
        writeArray_0(b_0);
    //               ^^^
    // When taking the "pointer syntax" path, valid WGSL is emitted:
    //  writeArray_0(&b_0);
        (&globalParams_outputBuffer_0)[dispatchThreadID_1.x] = i32(b_0[i32(0)].x);
        return;
    }

*Category 2 (tests/compute/inout)*

    fn testOut_0( x_0 : i32,  y_0 : ptr<function, i32>)
    {
    // Error while parsing WGSL: :5:5 error: cannot assign to parameter 'y_0'
        y_0 = x_0;
    //  ^^^
    // When taking the "pointer syntax" path, valid WGSL is emitted:
    //  *y_0 = x_0;
        return;
    }


*Category 3 (mutating-and-inout)*

    fn A_doThings_0( this_0 : ptr<function, A_0>,  y_0 : ptr<function, i32>)
    {
        var tmp_0 : i32 = this_0.x_0;
    // Error while parsing WGSL: :11:16 error: cannot assign 'ptr<function, i32, read_write>' to 'i32'
        this_0.x_0 = y_0;
    //             ^
    // When taking the "pointer syntax" path, valid WGSL is emitted:
    //  this_0.x_0 = *y_0;
    
        // ...
        
        return;
    }

Reply:
> I think we shouldn't emit things like `(&arr)[i]`...

In cases where arr is a structured buffer type, I currently have this hack (see the comment for my "explanation"):


    void WGSLSourceEmitter::emitOperandImpl(IRInst* inst, const EmitOpInfo& outerPrec)
    {
        // In WGSL, the structured buffer types are converted to ptr<AS, array<E>, AM>
        // everywhere, except for the global parameter declaration.
        // Thus, when these globals are used in expressions, we need an ampersand.
    
        if (inst->getOp() == kIROp_GlobalParam)
        {
            switch (inst->getDataType()->getOp())
            {
            case kIROp_HLSLStructuredBufferType:
            case kIROp_HLSLRWStructuredBufferType:
    
                m_writer->emit("(&");
                CLikeSourceEmitter::emitOperandImpl(inst, outerPrec);
                m_writer->emit(")");
                return;
            }
        }
    
        CLikeSourceEmitter::emitOperandImpl(inst, outerPrec);
    }

Should I solve this some other way, or insert some logic to not do this in the `structuredBuffer[i] = ...` case?

Reply:
I don't quite understand the comment here, what is `ptr<AS, array<E>, AM>` and why do we need `&` to access the elements in `array<E>`?

It feel like if HLSLStructuredBuffer needs to be mapped to a type with different level of indirection, then this shall be done in the legalization pass, with all use sites updated with the correct conversion/dereferencing insts so emit logic doesn't need to deal with this.

Reply:
> I don't quite understand the comment here, what is `ptr<AS, array<E>, AM>` and why do we need `&` to access the elements in `array<E>`?

The  `ptr<AS, T, AM>` (`T = array<E>` in this case) notation comes from https://www.w3.org/TR/WGSL/#ref-ptr-types.
AS = address space (e.g. storage, function, uniform, private, etc...), AM = access mode (R, RW).

We don't need `&` to access elements of `array`, but it doesn't hurt AFAIK.
(Array indexing notation like `buffer[i]` seems to work both when `buffer` is of type `ptr<storage, array<E>, read_write>` and just `array<E>`.)

For all other cases where a structured buffer is used as an operand (such as when it's passed to a function or assigned to a `ptr<AS, array<E>, AM>` typed thing, we need the ampersand because the global object corresponding to a structured buffer cannot be declared as a pointer type.

> It feel like if HLSLStructuredBuffer needs to be mapped to a type with different level of indirection, then this shall be done in the legalization pass, with all use sites updated with the correct conversion/dereferencing insts so emit logic doesn't need to deal with this.

Yeah, I think that there are two cases:

- When translating the global declaration of a structured buffer, `StructuredBuffer<E> buffer` should translate to `var<storage, read_write> buffer : array<E>`.
- Any other place, when referring to `buffer` the references should be translated to expressions of type `ptr<storage, array<E>, read_write>`.

This is essentially what I have done, but yes it feels fragile to do it at the emit stage.

Reply:
After the legalization, all references to structured buffers should be direct references to the global parameter, and there should not be any reference passing in the IR, so we shouldn't need to worry about case 2.